### PR TITLE
Apply suggestions from clang-tidy tool on array.cpp

### PR DESCRIPTION
**Pull request by umar456**
Clang-tidy is a tool for static analysis and other quality checks on software.
These changes are some of the suggestions from clang-tidy on the array.cpp file.
This commit does not add the clang-tidy config file because that is still a
work in progress.

### CODE REVIEW

File: src/api/cpp/array.cpp [hunk 1]
```diff
@@ -100,93 +135,71 @@ struct array::array_proxy::array_proxy_impl {
...
+        : arr(0) {                                                             \
+        af::dtype ty = static_cast<af::dtype>(dtype_traits<T>::af_type);       \
+        initDataArray(&arr, ptr, ty, src, dim0, dim1, dim2);                   \
+    }                                                                          \
+    template<>                                                                 \
+    AFAPI array::array(dim_t dim0, dim_t dim1, dim_t dim2, dim_t dim3,         \
+                       const T *ptr, af::source src)                           \
+        : arr(0) {                                                             \
+        af::dtype ty = static_cast<af::dtype>(dtype_traits<T>::af_type);       \
+        initDataArray(&arr, ptr, ty, src, dim0, dim1, dim2, dim3);             \
```
Comment (9prady9) [inline src/api/cpp/array.cpp]: `arr(0)` -> `arr(nullptr)` in all above template instantiations.

File: src/api/cpp/array.cpp [hunk 2]
```diff
@@ -100,93 +135,71 @@ struct array::array_proxy::array_proxy_impl {
...
 
-   private:
-    array_proxy_impl(const array_proxy_impl &);
-    array_proxy_impl(const array_proxy_impl &&);
-    array_proxy_impl operator=(const array_proxy_impl &);
-    array_proxy_impl operator=(const array_proxy_impl &&);
+    array_proxy_impl(const array_proxy_impl &)  = delete;
+    array_proxy_impl(const array_proxy_impl &&) = delete;
+    array_proxy_impl operator=(const array_proxy_impl &) = delete;
+    array_proxy_impl operator=(const array_proxy_impl &&) = delete;
```
Comment (9prady9) [inline src/api/cpp/array.cpp]: We should keep refactoring & style changes separate from the change like this. Perhaps a separate commit where the array_proxy_impl copy and move constructors are deleted.

Comment (umar456) [reply src/api/cpp/array.cpp]: This is equivalent to the previous code. Previously we made these constructors private which is the same as deleting them. Here we are using the `= delete` syntax.

**Comment by umar456** [comment]
Addressed all feedback

**Pull request merged**